dulwich/credentials.py

```diff
@@ -0,0 +1,122 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# You should have received a copy of the licenses; if not, see
+# <http://www.gnu.org/licenses/> for a copy of the GNU General Public License
+# and <http://www.apache.org/licenses/LICENSE-2.0> for a copy of the Apache
+# License, Version 2.0.
+#
+
+"""Support for git credential helpers
+
+https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage
+
+Currently Dulwich supports only the `get` operation
+
+"""
+import os
+import subprocess
+import sys
+from typing import Any, Dict, List, Optional, Union
+
+from dulwich.config import Config
+
+
+class CredentialNotFoundError(Exception):
+    pass
+
+
+class CredentialHelper:
+    """Helper for retrieving credentials for http/https git remotes
+
+    Usage:
+    >>> helper = CredentialHelper("store") # this example uses `git credential-store`
+    >>> credentials = helper.get("https://github.com/dtrifiro/aprivaterepo")
+    >>> username = credentials["username"]
+    >>> password = credentials["password"]
+    """
+
+    def __init__(self, command: str):
+        self.command = command
+        self._run_kwargs: Dict[str, Any] = {}
+        if self.command[0] == "!":
+            # On Windows this will only work in git-bash and/or WSL2
+            self._run_kwargs["shell"] = True
+
+    def _prepare_cmd(self) -> Union[str, List[str]]:
+        if self.command[0] == "!":
+            return self.command[1:]
+
+        argv = self.command.split(" ")
+        if not os.path.isabs(argv[0]):
+            args = " ".join(argv[1:])
+            cmd = f"git credential-{argv[0]}" + (f" {args}" if args else "")
+            return cmd.split(" ")
+
+        return self.command.split(" ")
+
+    def get(self, url: str) -> Dict[str, str]:
+        cmd = self._prepare_cmd()
+        if isinstance(cmd, str):
+            cmd = f"{cmd} get"
+        else:
+            cmd.append("get")
+
+        helper_input = f"url={url}\n".encode("UTF-8")
+
+        try:
+            res = subprocess.run(  # type: ignore # breaks on 3.6
+                cmd,
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                input=helper_input,
+                **self._run_kwargs,
+            )
+        except subprocess.CalledProcessError as exc:
+            raise CredentialNotFoundError(exc.stderr.decode("UTF-8")) from exc
+        except FileNotFoundError as exc:
+            raise CredentialNotFoundError("Helper not found") from exc
+
+        credentials = {}
+        for line in res.stdout.decode("UTF-8").strip().split("\n"):
+            try:
+                key, value = line.split("=")
+                credentials[key] = value
+            except ValueError:
+                continue
```

Comment:
I don't think we should be making assumptions about text encodings here, and just handle credentials as `Dict[bytes, bytes]` rather than `[str, str]`.

Text encoding for HTTP basic auth username/password is technically undefined (https://www.rfc-editor.org/rfc/rfc7617#appendix-B.3), and depending on the credential manager/system locale settings/etc, I think it's possible that the returned credentials here may not be utf-8 or ascii.

Then later we can just do `b64encode(username + b':' + password)` where the username/pw are already bytes objects.

Reply:
Agreed and corrected.

Note though that when username/password kwargs are provided to `Urllib3HttpGitClient()`, we use `urllib3.utils.make_headers`:
https://github.com/jelmer/dulwich/blob/bb83067f5d021b80d208dd1b93eda8d02578eeb2/dulwich/client.py#L2187

which encodes the `basic_auth` string as `latin-1`:
https://github.com/urllib3/urllib3/blob/972e9f02cd219892701fa0a1129dda7f81dac535/src/urllib3/util/request.py#L101-L104

In practice though, Imost of the time user/pass will be provided as kwargs when using `from_parsedurl` and in this case username/password are already decoded through `urlparse`.

Reply:
When using creds from a credential manager we should be generating the basic auth header ourselves. (We should be following the RFC, not urllib3)